### PR TITLE
Add `KindWindowsDesktops` to `ListResources`

### PR DESCRIPTION
**Pull request by kimlisa**
part of [RFD 55](https://github.com/gravitational/teleport/blob/master/rfd/0055-webui-ss-paginate-filter.md#filter-sort)

#### Description
To add support for type windows desktops to our generic `ListResources`, it needed it's own implementation `ListWindowsDesktops` in its own service layer (all other resources thus far belongs in the presence layer). This function is not made available to the client, they would need to use `ListResources` with type `KindWindowsDesktop`.

In addition adds support for sorting.

#### Required PR
- [x] Depends on this PR to have test pass: https://github.com/gravitational/predicate/pull/2

### CODE REVIEW

File: api/types/desktop.go [hunk 1]
```diff
@@ -218,3 +220,84 @@ func (f *WindowsDesktopFilter) Match(req WindowsDesktop) bool {
...
+
+// WindowsDesktops represents a list of windows desktops.
+type WindowsDesktops []WindowsDesktop
+
+// Len returns the slice length.
+func (s WindowsDesktops) Len() int { return len(s) }
+
+// Less compares desktops by name and host ID.
+func (s WindowsDesktops) Less(i, j int) bool {
+	return s[i].GetName() < s[j].GetName() && s[i].GetHostID() < s[j].GetHostID()
```
Comment (zmb3) [inline api/types/desktop.go:232]: This doesn't seem correct.

I think you want to sort by name, then host ID, so something like this:

```
switch {
case s[i].GetName() < s[j].GetName():
    return true
case s[i].GetName() > s[j].GetName():
    return false
default:
    return s[i].GetHostID() < s[j].GetHostID()  
}
```

File: api/types/desktop_test.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		{
+			name:      "by addr",
+			fieldName: ResourceSpecAddr,
+		},
+	}
+
+	for _, c := range cases {
+		c := c
+		t.Run(fmt.Sprintf("%s desc", c.name), func(t *testing.T) {
+			t.Parallel()
```
Comment (zmb3) [inline api/types/desktop_test.go:66]: You can remove the `t.Parallel()` from the subtests here. You only have two cases, and they are sorting tiny lists of 4 elements, so it's probably faster just to run them sequentially.

File: lib/auth/api.go [hunk 3]
```diff
@@ -809,6 +809,8 @@ type Cache interface {
 
 	// ListResources returns a paginated list of resources.
 	ListResources(ctx context.Context, req proto.ListResourcesRequest) (*types.ListResourcesResponse, error)
+	// ListWindowsDesktops returns a paginated list of windows desktops..
```
Comment (zmb3) [inline lib/auth/api.go:812]:
```suggestion
	// ListWindowsDesktops returns a paginated list of windows desktops.
```

File: lib/services/local/desktops_test.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+					PredicateExpression: `labels.env == "test"`,
+				},
+				expectedLen: 2,
+			},
+		}
+
+		for _, tc := range tests {
+			tc := tc
+			t.Run(tc.name, func(t *testing.T) {
+				t.Parallel()
```
Comment (zmb3) [inline lib/services/local/desktops_test.go:182]: I'm not a big fan of subtests of subtests of subtests. Can we simplify this and just have more top-level tests?

**Review by zmb3 — CHANGES_REQUESTED** [review]

File: api/types/desktop.go (hunk 1 above)
Comment (kimlisa) [reply api/types/desktop.go:232]: gah, didn't realize it was incorrect this entire time, sorry

File: lib/services/local/desktops_test.go (hunk 4 above)
Comment (kimlisa) [reply lib/services/local/desktops_test.go:182]: I separated the tests

**Review by espadolini — APPROVED** [review]
Is there a specific reason why desktops are not part of `Presence` like the other things? Should they be?

**Comment by kimlisa** [comment]
@espadolini it was b/c `KindWindowsDesktop` is not a keep alive type of resource, it belonged in the `services/local/desktop` (versus `KindDesktopWindowsService` that does belong in `services/local/presence`). The desktop resource works a bit differently from others in that a service does not hold information of the desktops it proxies.

File: api/types/desktop.go (hunk 1 above)
Comment (zmb3) [reply api/types/desktop.go]: No worries! It still would have sorted them in _some_ order, probably just not the order you'd expect :-)

**Review by rosstimothy — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

**Pull request merged**